python/ray/serve/utils.py

```diff
@@ -313,6 +314,53 @@ def parse_import_path(import_path: str):
     return ".".join(nodes[:-1]), nodes[-1]
 
 
+def merge_runtime_envs(parent_env: Dict, child_env: Dict) -> Dict:
```

Comment:
also should this be named override_runtime_envs_except_env_vars or something more accurate?

Reply:
I called it `merge_runtime_envs` since it implements [the current inheritance logic](https://docs.ray.io/en/releases-1.12.1/ray-core/handling-dependencies.html#inheritance). If this logic changes, we can update this function's implementation, which would automatically update all the callers.

We could call this `override_runtime_envs_except_env_vars`, but that name exposes implementation details. If the inheritance logic changes, we would need to make a new function and call it. The precision might be worth it though. What do you think?

Reply:
the thing is that it's no longer the "current inheritance logic" on Ray master anymore :(

Reply:
@simon-mo it is the current inheritance logic, since Ant's PR got reverted.  In the near future we'll decide on the final approach (whether to unrevert Ant's PR or not), and depending on that we might have to update this PR or make a new one if necessary

Reply:
If we decide to stick with this inheritance logic, in the future we'd like to move the inheritance logic out of C++ (where it currently is) and into the runtime env agent (Python).  Then we'll be able to call the same function in the runtime env agent and in this REST API code.

Reply:
@shrekris-anyscale what's the desired inheritance behavior you want in Serve again? Is it all merge based update, or override?

Reply:
Mainly, I want to make sure that deployments that don't have explicit `runtime_env`s defined can inherit the graph-level `runtime_env`. This ensures that their code can be deserialized when their replicas start, without requiring the user to copy-paste the graph environment to each deployment in the Serve config.

Other than that, I don't feel too strongly about the particular type of inheritance behavior, as long as it's clear and well-documented. As long as users understand how the graph-level environment and deployment-level environments interact, I think either merge-based or override-based inheritance is okay.

I chose to match the existing `runtime_env` inheritance behavior, so it would be less surprising for users that are already familiar with it. For example, users might read the existing inheritance documentation and simply assume that's how inheritance works in Serve as well, so I maintained consistency.

Reply:
@simon-mo What are your thoughts on merge-based vs override-based inheritance? Do you have another set of inheritance rules in mind?

Reply:
We synced offline. For now, we'll proceed with matching the existing `runtime_env` inheritance behavior and [introduce a test](https://github.com/ray-project/ray/pull/25502#issuecomment-1149247529) to catch any deviation between the general inheritance behavior and Serve's inheritance behavior. We'll rename `merge_runtime_envs` to `override_runtime_envs_except_env_vars`.